Document `#[repr(u8)] enum` to enable LLVM optimizations

geofft opened the pull request:
See rust-lang/rust#27303 and rust-lang/libc#48 for discussion of the
optimization, or just try compiling `unsafe {free(malloc(16))}` with various
Rust definitions of `void *` in the FFI binding.

Gankro commented:
This is a bit unclear to me as written. I would maybe zoom out from LLVM a bit and focus on the ABI implications of C-like enums: it will be treated as an integer, and not a struct-containing-an-integer.

CC @mystor

geofft commented:
Oh, hm, is a `#[repr(u8)] enum Foo {Bar(())}` represented as a struct, and thus passed on the stack in ABIs where structs are unconditionally on the stack? Then yeah, that's worth documenting too! Let me test that and then rework this paragraph, thanks.

mystor commented:
It might also be good to note that controlling the `#[repr]` of the enum doesn't change the set of legal values, and that unlike C/C++ enums, it is undefined behaviour to have a `#[repr(C)] enum Foo { A = 0, B = 1 }` with the value `10`.

geofft commented:
OK, I've rewritten it to take the above comments into account.

This has drifted a fair bit from my original subject; the specific optimization LLVM does with `malloc` and `free` (and particular, that LLVM considers `void *` to be a `* u8`) is more than I can fit in a passing comment. However, the comments on `std::os::raw::c_void` and `libc::c_void`, _plus_ the paragraph here about the enum needing to be C-like, are probably clear enough.

ubsan commented:
> If a C-like enum is missing a `repr(u*)` or `repr(C)`, it is treated like a one-member struct for ABI compatibility purposes.

This seems like a silly guarantee to make, since they should *not* be treated as a one-member struct for ABI purposes.

steveklabnik commented:
I have no idea what's right here...

mystor commented:
I generally agree with @ubsan That any properties of `repr(rust)` enums, with the potential exception of `Option<&T>` and `Option<Box<T>>`, should not be documented. Any internal implementation details of them should remain internal.

Gankro commented:
Agreed, the only property of repr(rust) that's guaranteed is that `OptionLike<&T>` has the same abi as `&T` and `*const T`.

steveklabnik commented:
Great. @geofft care to make these changes?

geofft commented:
Oh yeah, good points about not documenting the exact properties of `repr(rust)`. Thanks!

I've updated the pull request. This doesn't seem like the right area to discuss `OptionLike`, so I just removed all discussion of `repr(rust)` representation, but I can add a bit about that if that's helpful.

steveklabnik commented:
Let's :shipit: . Thanks!
